### PR TITLE
basic-install with openSUSE support

### LINKED ISSUE #4947: Installer dies silently if OS is not supported
### Versions

- Pi-hole: 5.12.2 
- AdminLTE:  
- FTL:  5.18.1

### Platform

- OS and version:  openSUSE MicroOS 20220921

- Platform:  Raspberry Pi 3

### Expected behavior

The installation script tells the user what the issue is, resp. asks for input like package manager to use etc

### Actual behavior / bug

The installation script determines that the underlying system uses RPM as package manager, however, it searches for some redhat files (see debug output) which it does not find and just finished. No output to the user

### Steps to reproduce

According to https://docs.pi-hole.net/main/basic-install/ use alternative 2

No further steps required

## Debug Token

see output of installation script (executed with set -x)

### Screenshots

_If applicable, add screenshots to help explain your problem._

### Additional context

[install.log](https://github.com/pi-hole/pi-hole/files/9657876/install.log)

### PR DESCRIPTION
**Pull request by coogor**
replacing PR #5000 against master

Signed-off-by: coogor <axel.braun@gmx.de>

## Thank you for your contribution to the Pi-hole Community! 

Please read the comments below to help us consider your Pull Request.

We are all volunteers and completing the process outlined will help us review your commits quicker.

**Please make sure you**
 1. Base your code and PRs against the repositories developmental branch.
 2. [Sign Off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits as we enforce the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions

---
- **What does this PR aim to accomplish?:**

This PR adds support for openSUSE as discussed in #4947

- **How does this PR accomplish the above?:**
- 
It adds determination for zypper as package manager. It adds packages (naming) used in openSUSE to achieve proper installation of the pihole package. Is uses systemd-service (as soon as available)

- **What documentation changes (if any) are needed to support this PR?:**

Comments and messages with relation to openSUSE are added to the installation script.
Installer was tested on openSUSE Tumbleweed and MicroOS

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code and I have tested my changes.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

---
- [x] I have read the above and my PR is ready for review. _Check this box to confirm_

### CODE REVIEW

**Comment by dschaper** [comment]
Thank you, could you please fill out the template for this PR. The template will be the documentation we refer to if this PR is merged and having the information there is very helpful when we need to refer back to what this PR was meant to do.

**Comment by coogor** [comment]
**What does this PR aim to accomplish?:**

This PR adds support for openSUSE as discussed in https://github.com/pi-hole/pi-hole/issues/4947

    **How does this PR accomplish the above?:**

It adds determination for zypper as package manager. It  adds packages (naming) used in openSUSE to achieve proper installation of the pihole package. Is uses systemd-service (s soon as available)

    **What documentation changes (if any) are needed to support this PR?:**

Comments and messages with relation to openSUSE are added to the installation script.
Installer was tested on openSUSE Tumbleweed and MicroOS

**Comment by coogor** [comment]
> Thank you, could you please fill out the template for this PR. The template will be the documentation we refer to if this PR is merged and having the information there is very helpful when we need to refer back to what this PR was meant to do.

Template was filled, let me know if you need adidtional information.
I have seen some check report failure (trailing whitespace) - Can I somehow update the submission/pull request or do I need to create a new one? Sry, not familiar with github!

**Comment by yubiuser** [comment]
You can make your modifications locally, commit them locally and push to your fork on github. This PR will automatically update.

**Comment by rdwebdesign** [comment]
@coogor 

> Template was filled, let me know if you need adidtional information.

You need to **Edit** the PR text.
Use the options menu, select Edit and change the original post.

image

File: automated install/basic-install.sh [hunk 1]
```diff
@@ -2638,6 +2662,8 @@ main() {
         fi
     fi
 
+    echo "password is " ${pw}
```
Comment (rdwebdesign) [inline automated install/basic-install.sh:2665]: You need to use double quotes around the variable here.

```suggestion
    echo "password is " "${pw}"
```

or

```suggestion
    echo "password is ${pw}"
```

**Review by rdwebdesign — CHANGES_REQUESTED** [review]

**Comment by coogor** [comment]
Edited the description and added double quotes

File: automated install/basic-install.sh [hunk 2]
```diff
@@ -351,37 +351,59 @@ package_manager_detect() {
         # Then check if dnf or yum is the package manager
         if is_command dnf ; then
             PKG_MANAGER="dnf"
-        else
+        fi
```
Comment (yubiuser) [inline automated install/basic-install.sh:354]: I think you could go with an `elif` here.

**Review by yubiuser — COMMENTED** [review]
The whole section L361 - L405 has a lot of duplicated code. Maybe you could separate what they OS have in common and only apply a `if [ OS = XX]` when really needed.

File: automated install/basic-install.sh [hunk 3]
```diff
@@ -2625,6 +2647,8 @@ main() {
     # Copy the temp log file into final log location for storage
     copy_to_install_log
 
+    echo 'pasword generation'
```
Comment (yubiuser) [inline automated install/basic-install.sh:2650]: Why do you need this?

Also L 2665, L 2722, L2731, L2738?

File: automated install/basic-install.sh (hunk 2 above)
Comment (coogor) [reply automated install/basic-install.sh]: Correct, changed

File: automated install/basic-install.sh (hunk 3 above)
Comment (coogor) [reply automated install/basic-install.sh:2649]: Well, these are not really necessary, but helped understanding the program flow, resp, where it got stuck.
Do you think these should be removed?

Comment (rdwebdesign) [reply automated install/basic-install.sh:2649]: > but helped understanding the program flow

I understand you used this to "debug" each step while you where learning each code step, but i think these messages are not needed on the installer.

Please remove them.

**Review by yubiuser — COMMENTED** [review]
I refactored your code a bit to reduce doubling. What do you think?

```
      # These variable names match the ones for apt-get. See above for an explanation of what they are for.
      PKG_INSTALL=("${PKG_MANAGER}" install -y)
      OS_CHECK_DEPS=(grep bind-utils)
      LIGHTTPD_USER="lighttpd"
      LIGHTTPD_GROUP="lighttpd"
      LIGHTTPD_CFG="lighttpd.conf.fedora"
        
        if [ "${PKG_MANAGER}" = "dnf" ] || [ "${PKG_MANAGER}" = "yum" ]  then
          # CentOS package manager returns 100 when there are packages to update so we need to || true to prevent the script from exiting.
          PKG_COUNT="${PKG_MANAGER} check-update | egrep '(.i686|.x86|.noarch|.arm|.src)' | wc -l || true"
          INSTALLER_DEPS=(git dialog iproute newt procps-ng which chkconfig ca-certificates)
          PIHOLE_DEPS=(cronie curl findutils sudo unzip libidn2 psmisc libcap nmap-ncat jq)
          PIHOLE_WEB_DEPS=(lighttpd lighttpd-fastcgi php-common php-cli php-pdo php-xml php-json php-intl)

        # If the host OS is centos (or a derivative), epel is required for lighttpd
          if ! grep -qiE 'fedora|fedberry' /etc/redhat-release; then
              if rpm -qa | grep -qi 'epel'; then
                  printf "  %b EPEL repository already installed\\n" "${TICK}"
              else
                  local RH_RELEASE EPEL_PKG
                  # EPEL not already installed, add it based on the release version
                  RH_RELEASE=$(grep -oP '(?<= )[0-9]+(?=\.?)' /etc/redhat-release)
                  EPEL_PKG="https://dl.fedoraproject.org/pub/epel/epel-release-latest-${RH_RELEASE}.noarch.rpm"
                  printf "  %b Enabling EPEL package repository (https://fedoraproject.org/wiki/EPEL)\\n" "${INFO}"
                  "${PKG_INSTALL[@]}" "${EPEL_PKG}"
                  printf "  %b Installed %s\\n" "${TICK}" "${EPEL_PKG}"
              fi
          fi
        else
          # openSUSE packages
          # openSUSE package manager returns number when there are packages to update so we need to || true to prevent the script from exiting.
          PKG_COUNT="${PKG_MANAGER} lu | egrep '(.i686|.x86|.noarch|.arm|.src)' | wc -l || true"
          INSTALLER_DEPS=(git dialog iproute newt procps which ca-certificates)
          PIHOLE_DEPS=(cronie curl findutils sudo unzip libidn2 psmisc libcap-ng0 nmap ncat jq)
          PIHOLE_WEB_DEPS=(lighttpd php8 php8-fastcgi php8-cli php8-pdo php8-intl php8-openssl php8-sqlite)
```

**Comment by yubiuser** [comment]
I setup a test environment for openSUSE Thumbleweed to be able to test your PR. The test will pull the latest image of opensuse/tumbleweed and run our test suite against your changes.

As you can see, it fails at some tests:
https://github.com/pi-hole/pi-hole/actions/runs/3500340861/jobs/5862908143

To continue, please provide the output of `cat /etc/os-release` for the current thumbleweed

**Comment by yubiuser** [comment]
Also: please rebase on latest development and sovle the merge conflicts.

**Comment by coogor** [comment]
Thank you, I wanted to have a look into this on the weekend, but dont mind you were faster :-)
Rebasing...looks like this does not work automatically. 'Classic' approach with editor I guess?

**Comment by yubiuser** [comment]
Yes, you need to rebase locally and force push afterwards.

**Comment by coogor** [comment]
> I setup a test environment for openSUSE Thumbleweed to be able to test your PR. The test will pull the latest image of opensuse/tumbleweed and run our test suite against your changes.
> 
> As you can see, it fails at some tests: https://github.com/pi-hole/pi-hole/actions/runs/3500340861/jobs/5862908143
> 
> To continue, please provide the output of `cat /etc/os-release` for the current thumbleweed

docb@X1E:~/buildservice> cat /etc/os-release 
NAME="openSUSE Tumbleweed"
# VERSION="20221115"
ID="opensuse-tumbleweed"
ID_LIKE="opensuse suse"
VERSION_ID="20221115"
PRETTY_NAME="openSUSE Tumbleweed"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:opensuse:tumbleweed:20221115"
BUG_REPORT_URL="https://bugs.opensuse.org"
HOME_URL="https://www.opensuse.org/"
DOCUMENTATION_URL="https://en.opensuse.org/Portal:Tumbleweed"
LOGO="distributor-logo-Tumbleweed"

**Comment by yubiuser** [comment]
Tumbleweed, as a rolling distribution, has only a `latest` image we could test against. A often changing image is not a good foundation for tests. I will change the test image therefore to Leap 15.4.

**Comment by coogor** [comment]
oops, why was that closed.
rebase has happened, against original extentions. 
No trying to apply refactored code

**Comment by yubiuser** [comment]
Tests are currently failing at

`/opt/pihole/basic-install.sh: line 2284: which: command not found\n').stdout`

However, we don't use `which` at that place anymore:
https://github.com/pi-hole/pi-hole/blob/e47f4180f35e88239e974c147bbab43c95e23566/automated%20install/basic-install.sh#L2284

This was changed in https://github.com/pi-hole/pi-hole/pull/4907
__

This makes me think, something might went wrong with your rebase. Did you pull the `development` branch before rebasing?

**Comment by yubiuser** [comment]
https://github.com/pi-hole/pi-hole/blob/b2ec27ce3996e289eb59bbed16f93b3011ff381a/automated%20install/basic-install.sh#L2284

**Comment by coogor** [comment]
> Did you pull the `development` branch before rebasing?
Yes.....

> That makes me think, something might went wrong with your rebase. Did you pull the `development` branch before rebasing?

I Think the commit line fornthis file mentioned something of 'which' in it.
What do you propose to continue? Sorry for causing work, this is the first time I really work on github

**Comment by yubiuser** [comment]
> Sorry for causing work, this is the first time I really work on github

No worries. We all started somewhere and made the same mistakes. We are here to help. 

> What do you propose to continue? 

You would need to refresh (pull) the latest development changes from `pi-hole` into your local fork and rebase again. This might cause merge conflicts now. 
If you feel like this is to much trouble (but bonus: you learn something about `git`), you could start the PR from scratch (not a big deal, you can basically copy&paste) your changes as they all happen in one file so far. However, as I'm working on the test suite in parallel, you would need to update your development branch and rebase anyway, as we would merge the test suite first and then this PR (to really test this PR before merging).

**Comment by coogor** [comment]
> > What do you propose to continue?
> 
> You would need to refresh (pull) the latest development changes from `pi-hole` into your local fork and rebase again. This might cause merge conflicts now. 

No...it says I'm on the latest stage: This branch is not behind the upstream pi-hole:development  (3 commits ahead)

If you feel like this is to much trouble (but bonus: you learn something about `git`), you could start the PR from scratch (not a big deal, you can basically copy&paste) your changes as they all happen in one file so far. However, as I'm working on the test suite in parallel, you would need to update your development branch and rebase anyway, as we would merge the test suite first and then this PR (to really test this PR before merging).

**Comment by coogor** [comment]
Tested the latest version of installation script successfully up to the point where pihole-ftl.service should be invoked.
AFAIK the proper systemd-service is not in the master branch yet?

**Comment by yubiuser** [comment]
No it's not. I'm working on the test suite, but it needs some more work

**Comment by yubiuser** [comment]
Tests are passing now. 

https://github.com/pi-hole/pi-hole/actions/runs/3527352033/jobs/5938401382

However, I discovered a few places in `basic-install` that also need changes (because there are conditions we do not capture atm, which where revealed by testing Leap 15). I'll open some PRs (think of 3-4) to fix them. When they all are merged, you need to rebase your PR on `development` again to be able to pass the Leap test suite.

**Comment by coogor** [comment]
> Tests are passing now.
> 
> https://github.com/pi-hole/pi-hole/actions/runs/3527352033/jobs/5938401382

good. However, I feel Tumbleweed would be the better test target, as it is the basis for MicroOS (small, transactional, self-maintaining...) and this is IMO the best solution for a local DNS server...
 
> However, I discovered a few places in `basic-install` that also need changes (because there are conditions we do not capture atm, which where revealed by testing Leap 15). I'll open some PRs (think of 3-4) to fix them. When they all are merged, you need to rebase your PR on `development` again to be able to pass the Leap test suite.

OK, let me know once done. I had just synced the devel branch and that was OK. Or create the PR directly against my fork :-)

**Comment by rdwebdesign** [comment]
Tumbleweed is a rolling release and we probably will only support stable releases.

**Comment by yubiuser** [comment]
https://github.com/pi-hole/pi-hole/pull/5038
https://github.com/pi-hole/pi-hole/pull/5039

File: automated install/basic-install.sh [hunk 4]
```diff
@@ -351,40 +351,51 @@ package_manager_detect() {
...
+                  "${PKG_INSTALL[@]}" "${EPEL_PKG}"
+                  printf "  %b Installed %s\\n" "${TICK}" "${EPEL_PKG}"
+              fi
+          fi
+        else
+          # openSUSE packages
+          # openSUSE package manager returns number when there are packages to update so we need to || true to prevent the script from exiting.
+          PKG_COUNT="${PKG_MANAGER} lu | egrep '(.i686|.x86|.noarch|.arm|.src)' | wc -l || true"
+          INSTALLER_DEPS=(git dialog iproute newt procps which ca-certificates)
+          PIHOLE_DEPS=(cronie curl findutils sudo unzip libidn2 psmisc libcap-ng0 nmap ncat jq)
```
Comment (yubiuser) [inline automated install/basic-install.sh:393]:
```suggestion
          PIHOLE_DEPS=(cronie curl findutils sudo unzip libidn2 psmisc libcap-ng0 netcat-openbsd jq)
```

File: automated install/basic-install.sh [hunk 5]
```diff
@@ -351,40 +351,51 @@ package_manager_detect() {
...
+                  printf "  %b Enabling EPEL package repository (https://fedoraproject.org/wiki/EPEL)\\n" "${INFO}"
+                  "${PKG_INSTALL[@]}" "${EPEL_PKG}"
+                  printf "  %b Installed %s\\n" "${TICK}" "${EPEL_PKG}"
+              fi
+          fi
+        else
+          # openSUSE packages
+          # openSUSE package manager returns number when there are packages to update so we need to || true to prevent the script from exiting.
+          PKG_COUNT="${PKG_MANAGER} lu | egrep '(.i686|.x86|.noarch|.arm|.src)' | wc -l || true"
+          INSTALLER_DEPS=(git dialog iproute newt procps which ca-certificates)
```
Comment (yubiuser) [inline automated install/basic-install.sh:392]:
```suggestion
          INSTALLER_DEPS=(git dialog iproute2 newt procps which ca-certificates)
```

File: automated install/basic-install.sh [hunk 6]
```diff
@@ -351,40 +351,51 @@ package_manager_detect() {
...
+          PIHOLE_DEPS=(cronie curl findutils sudo unzip libidn2 psmisc libcap-ng0 nmap ncat jq)
+          PIHOLE_WEB_DEPS=(lighttpd php8 php8-fastcgi php8-cli php8-pdo php8-intl php8-openssl php8-sqlite)
```
Comment (yubiuser) [inline automated install/basic-install.sh:394]:
```suggestion
          PIHOLE_WEB_DEPS=(lighttpd php php-fastcgi php-cli php-pdo php-intl php-openssl php-sqlite)
```

File: automated install/basic-install.sh (hunk 5 above)
Comment (yubiuser) [reply automated install/basic-install.sh:392]: `iproute` is replaced by `iproute2`

File: automated install/basic-install.sh (hunk 4 above)
Comment (yubiuser) [reply automated install/basic-install.sh:393]: Leap does not have `nmap` as separate package, `netcat-openbsd` is available for `tumbleweed` and `leap`

File: automated install/basic-install.sh (hunk 6 above)
Comment (yubiuser) [reply automated install/basic-install.sh:394]: Please do not hard-code a `php` version. E.g. `leap` on `x86_64` has `php8` while on `arm` `php7.4 is the default.

**Comment by yubiuser** [comment]
Test on `tumbleweed` fail until https://github.com/pi-hole/pi-hole/pull/4924/ is merged to. Adding this to the test suit and adjusting the test:
https://github.com/pi-hole/pi-hole/actions/runs/3550794823

(It fails for tumbleweed because it's not an supported OS, all other tests pass)

**Comment by yubiuser** [comment]
Tumbleweed passes now as well: https://github.com/pi-hole/pi-hole/actions/runs/3553020655/jobs/5977810768

File: automated install/basic-install.sh [hunk 7]
```diff
@@ -351,40 +351,51 @@ package_manager_detect() {
         # Then check if dnf or yum is the package manager
         if is_command dnf ; then
             PKG_MANAGER="dnf"
-        else
+        elif is_command yum ; then
             PKG_MANAGER="yum"
+        elif is_command zypper ; then
```
Comment (yubiuser) [inline automated install/basic-install.sh:356]: Please see this comment

https://github.com/pi-hole/pi-hole/pull/5027#discussion_r1046585261

**Comment by coogor** [comment]
every time I sync with devel, this gets closed

File: automated install/basic-install.sh [hunk 8]
```diff
@@ -1385,6 +1395,11 @@ installConfigs() {
     if [[ -d '/run/systemd/system' ]]; then
         install -T -m 0644 "${PI_HOLE_LOCAL_REPO}/advanced/Templates/pihole-FTL.systemd" '/etc/systemd/system/pihole-FTL.service'
 
+        # Set net admin permissions so that FTL can serve DNS, DHCP and IMAP (for DHCPv6). If this does not work, run FTL as root user.
```
Comment (yubiuser) [inline automated install/basic-install.sh:1398]: Did you re-add this intentionally? It was removed by https://github.com/pi-hole/pi-hole/pull/5043

**Comment by coogor** [comment]
No. Probably the mess with rebasing / merging due to new changes in development branch. Can you remove it again? Thx!
Viele Grüße
Axel 
-- 
Diese Nachricht wurde von meinem Android-Tablet mit K-9 Mail gesendet.

Am 21. Dezember 2022 00:23:34 MEZ schrieb yubiuser ***@***.***>:

**Comment by yubiuser** [comment]
I cherry-picked your commit (to preserve your authorship) and removed the additional lines. I also combined it with my PR to implement the necessary test suites. See #5083 5083

**Comment by coogor** [comment]
Thanks! looks like go-live gets closer :-)

**Comment by coogor** [comment]
can we close this pull request?

**Pull request closed without merging**